### PR TITLE
Add option to enable token registration without requiring 3pids

### PR DESCRIPTION
**Pull request by Half-Shot**
This PR attempts to add a second flow for token registration which does not require 3pid validation to succeed, *regardless* of the configuration of `registrations_require_3pid`. The requirement is that we would like to be able to offer tokens that allow users to sign up for a homeserver without having to jump through a email / msidn verification flow.

I'm not a massive fan of the config flag that I have come up with, and I'm a little worried this feels like adding another layer upon an already complicated configuration system (and at this point, does it just become easier to allow advanced users to craft their own registration flows?). That said, I think the idea of having temporal tokens that allow users to bypass registration requirements could be a useful thing, so it's worth figuring out the best configuration option for it.

### CODE REVIEW

**Comment by clokep** [comment]
This feels a little bit like it weakens #12091, but i guess the user still needs a registration token so maybe not?

**Comment by Half-Shot** [comment]
> This feels a little bit like it weakens #12091, but i guess the user still needs a registration token so maybe not?

Yup, the implication is that you aren't sharing your registration token publicly if you enable this option (or in our module-enhanced case, the shared token doesn't let you do much, so it's of little value).

**Review by squahtx — COMMENTED** [review]
Not sure I fully understand the new config option.

I understand it adds an alternative flow, requiring a registration token, that is an alternative to the required 3pids?
And the new config option only really makes sense if `registration_requires_token` is off, otherwise you could clear out `registrations_require_3pid`?

**Comment by Half-Shot** [comment]
> Not sure I fully understand the new config option.
> 
> I understand it adds an alternative flow, requiring a registration token, that is an alternative to the required 3pids? And the new config option only really makes sense if `registration_requires_token` is off, otherwise you could clear out `registrations_require_3pid`?

Yeah, so I suppose this is best described as either the requestor must provide a 3pid XOR a registration token. The key here is we don't want to require a token if someone provides 3pid, and we don't want to require a 3pid if they provide a token.

It's annoyingly a bit of an edge case, and it's just not really possible to configure Synapse to work that way without another config field. It's a bit of a shame we couldn't extend `registrations_require_3pid` to include registration tokens as a valid option..

File: synapse/config/registration.py [hunk 1]
```diff
@@ -309,6 +312,16 @@ def generate_config_section(
...
 
+        # Allow users to submit a token during registration without requiring them to complete any
+        # 3pid steps.
+        # Tokens can be managed using the admin API:
+        # https://matrix-org.github.io/synapse/latest/usage/administration/admin_api/registration_tokens.html
+        # Note that `enable_registration` must be set to `true`.
+        # Disabling this option will not delete any tokens previously generated.
+        # Defaults to false. Uncomment the following to require tokens:
+        #
+        #registration_token_without_3pid: false
```
[the remark this answers is not in the document]
Comment (Half-Shot) [reply synapse/config/registration.py:323]: I'm okay with that.

File: synapse/rest/client/register.py [hunk 2]
```diff
@@ -929,6 +929,15 @@ def _calculate_registration_flows(
         # always let users provide both MSISDN & email
         flows.append([LoginType.MSISDN, LoginType.EMAIL_IDENTITY])
 
+    # Prepend registration token to all flows if we're requiring a token
+    if config.registration.registration_requires_token:
+        for flow in flows:
+            flow.insert(0, LoginType.REGISTRATION_TOKEN)
```
[the remark this answers is not in the document]
Comment (Half-Shot) [reply synapse/rest/client/register.py:935]: Oh likely yes, let me rethink this.

File: synapse/rest/client/register.py [hunk 3]
```diff
@@ -942,7 +946,8 @@ def _calculate_registration_flows(
     # Prepend registration token to all flows if we're requiring a token
     if config.registration.registration_requires_token:
         for flow in flows:
-            flow.insert(0, LoginType.REGISTRATION_TOKEN)
+            if LoginType.REGISTRATION_TOKEN not in flow:
```
Comment (Half-Shot) [inline synapse/rest/client/register.py:949]: I've gone for this check, which isn't as nice but it does preserve the order.

File: synapse/rest/client/register.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Half-Shot) [reply synapse/rest/client/register.py]: I've changed this to behave like before.

**Review by squahtx — APPROVED** [review]
Thanks for making the changes! LGTM!

File: synapse/rest/client/register.py [hunk 4]
```diff
@@ -929,6 +929,10 @@ def _calculate_registration_flows(
         # always let users provide both MSISDN & email
         flows.append([LoginType.MSISDN, LoginType.EMAIL_IDENTITY])
 
+    # Add a flow that doesn't require any 3pids, if the config requests it.
+    if config.registration.enable_registration_token_3pid_bypasss:
+        flows.append([LoginType.REGISTRATION_TOKEN])
```
Comment (buffless-matt) [inline synapse/rest/client/register.py:934]: Should this have been:
```
flows.append([LoginType.REGISTRATION_TOKEN, LoginType.DUMMY])
```

...for similar reasons as stated [here](https://github.com/matrix-org/synapse/blob/78b99de7c206b106340e12cdee0af9aa246bd5ad/synapse/rest/client/register.py#L914-L917)?

**Pull request merged**